### PR TITLE
br-stream: use raft router to apply kv files for sst_importer

### PR DESCRIPTION
**Pull request by 3pointer**
### What is changed and how it works?

What's Changed:
Use raft router to apply kv changed file instead of engine.

```
br-stream: use raft router to apply kv files for sst_importer
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
1. create table and do a full backup
2. starts log backup with tpcc bench
3. check the data
image
4. do a full restore and log restore with this PR
5. check the data
image
6. check the log
`[2022/02/24 14:24:26.757 +08:00] [DEBUG] [sst_service.rs:525] ["finished apply kv file with Ok(range { start: 7480000000000000FFAC5F720380000000FF0000000103800000FF0000000001038000FF0000000000010000FDFA0358980EEBFFFE end: 7480000000000000FFAC5F720380000000FF0000000103800000FF000000000A038000FF000000000BB80000FDFA0358992F2FFFFE })"]`

Side effects

- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/import/sst_service.rs [hunk 1]
```diff
@@ -90,6 +94,7 @@ where
             importer,
             limiter: Limiter::new(f64::INFINITY),
             task_slots: Arc::new(Mutex::new(HashSet::default())),
+            raft_entry_max_size: RaftConfig::default().raft_entry_max_size,
```
Comment (YuJuncen) [inline src/import/sst_service.rs]: Maybe make it an argument and get it from the real config?

File: src/import/sst_service.rs [hunk 2]
```diff
@@ -374,40 +430,96 @@ where
...
+                            let (cb, future) = paired_future_callback();
+                            match router.send_command(
+                                cmd,
+                                Callback::write(cb),
+                                RaftCmdExtraOpts::default(),
+                            ) {
+                                Ok(_) => futs.push(future),
+                                Err(_e) => {
+                                    let mut import_err = kvproto::import_sstpb::Error::default();
+                                    import_err
```
Comment (YuJuncen) [inline src/import/sst_service.rs]: Maybe provide the context(`_e`)?

File: src/import/sst_service.rs [hunk 3]
```diff
@@ -374,40 +430,96 @@ where
...
-            let mut resp = ApplyResponse::default();
-            match res {
-                Ok(range) => {
-                    if let Some(r) = range {
-                        resp.set_range(r);
+            let resp = if !futs.is_empty() {
+                Ok(join_all(futs).await.iter().fold(apply_resp, |mut resp, x| {
+                    if x.is_err() {
+                        let mut import_err = kvproto::import_sstpb::Error::default();
+                        import_err.set_message("failed to complete raft command".to_string());
```
Comment (YuJuncen) [inline src/import/sst_service.rs]: Maybe collect those errors and send back to BR?

File: src/import/sst_service.rs [hunk 4]
```diff
@@ -155,6 +160,57 @@ where
         None
     }
 
+    fn build_apply_request<'a, 'b>(
+        &self,
+        reqs: &'a mut Vec<Request>,
+        cmd_reqs: &'a mut Vec<RaftCmdRequest>,
+        is_delete: bool,
+        cf: &'b str,
+        context: Context,
```
Comment (YuJuncen) [inline src/import/sst_service.rs:170]: `Context` cloud be a reference, or perhaps we can remove the `clone` in L190 and L207?

File: src/import/sst_service.rs [hunk 5]
```diff
@@ -374,40 +430,101 @@ where
...
                     }
-                }
-                Err(e) => resp.set_error(e.into()),
-            }
-            let resp = Ok(resp);
+                    resp
+                }))
+            } else {
+                Ok(apply_resp)
+            };
```
Comment (YuJuncen) [inline src/import/sst_service.rs]: Considering `[].fold(apply_resp, _) == apply_resp`, maybe the `if-else` can be removed?

**Review by YuJuncen — APPROVED** [review]
rest lgtm

File: src/import/sst_service.rs [hunk 6]
```diff
@@ -374,40 +430,101 @@ where
...
+                        for cmd in cmd_reqs {
+                            let (cb, future) = paired_future_callback();
+                            match router.send_command(
+                                cmd,
+                                Callback::write(cb),
+                                RaftCmdExtraOpts::default(),
+                            ) {
+                                Ok(_) => futs.push(future),
+                                Err(_e) => {
+                                    let mut import_err = kvproto::import_sstpb::Error::default();
```
Comment (kennytm) [inline src/import/sst_service.rs]: this is nested too deeply. can you do something like:

```rust
let result = (|| -> Result<()> {
    let temp_file = importer.do_download_kv_file(...)?;
    // etc.
    Ok(())
})();
if let Err(e) = result {
    apply_resp.set_error(e.into());
}
```

File: src/import/sst_service.rs (hunk 4 above)
Comment (3pointer) [reply src/import/sst_service.rs:170]: if `Context` become a reference, then `make_request_header` need another clone for Context. 🤔

Comment (YuJuncen) [reply src/import/sst_service.rs:170]: So maybe just remove those clones? (will rustc complain with that?)

File: src/import/sst_service.rs (hunk 2 above)
Comment (3pointer) [reply src/import/sst_service.rs]: fixed.

File: src/import/sst_service.rs (hunk 1 above)
Comment (3pointer) [reply src/import/sst_service.rs]: fixed.

File: src/import/sst_service.rs (hunk 5 above)
Comment (3pointer) [reply src/import/sst_service.rs]: fixed

File: src/import/sst_service.rs (hunk 6 above)
Comment (3pointer) [reply src/import/sst_service.rs]: Done

File: src/import/sst_service.rs (hunk 4 above)
Comment (3pointer) [reply src/import/sst_service.rs:170]: since closure is FnMut，so we can not just remove those clones.🤔

File: src/import/sst_service.rs (hunk 3 above)
Comment (3pointer) [reply src/import/sst_service.rs]: fixed

File: src/import/sst_service.rs [hunk 7]
```diff
@@ -747,3 +851,11 @@ fn make_request_header(mut context: Context) -> RaftRequestHeader {
     header.set_region_epoch(context.take_region_epoch());
     header
 }
+
+fn make_request(reqs: &mut Vec<Request>, context: Context) -> RaftCmdRequest {
+    let mut cmd = RaftCmdRequest::default();
+    let header = make_request_header(context);
+    cmd.set_header(header);
+    cmd.set_requests(reqs.drain(..).collect());
```
Comment (kennytm) [inline src/import/sst_service.rs]:
```suggestion
    cmd.set_requests(mem::take(reqs));
```

**Review by kennytm — APPROVED** [review]
rest lgtm

**Pull request merged**